### PR TITLE
Fedora mingw

### PR DESCRIPTION
**Pull request by elpaso**
Replace MXE with fedora mingw64 cross build recipe with python.

Thanks to Sandro Mani!

## TODO

- [x] - ccache

### CODE REVIEW

**Comment by elpaso** [comment]
CC @manisandro first test run. ccache support is still missing

**Comment by roya0045** [comment]
Hopefully this option will be less of a hassle for you than MXE. 👍

File: .github/workflows/minw64.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+jobs:
+  build:
+    name: MinGW64 Windows Build
+    runs-on: ubuntu-latest
+    steps:
+
+      - uses: actions/checkout@v1
+      - name: Build Docker Container with Build Environment
+        run: docker build -t qgis3-mingw-buildenv -f ms-windows/mingw/qgis3-build-deps-mingw.dockerfile .
```
Comment (3nids) [inline .github/workflows/minw64.yml:38]: this needs to be cached. You can use a hashkey on the the dockerfile as part of the key

File: .github/workflows/minw64.yml [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+      - name: Build QGIS Application
+        run: docker run -w /QGIS -v $PWD:/QGIS qgis3-mingw-buildenv ms-windows/mingw/build.sh x86_64 nodebug 4
```
Comment (3nids) [inline .github/workflows/minw64.yml:41]: can you add ccache similarly to what has been done in 4fc02733cb7bdf517eb5845f2883e04ef2e6f428

File: .github/workflows/minw64.yml (hunk 1 above)
Comment (elpaso) [reply .github/workflows/minw64.yml:37]: @3nids mind having a look if I implemented the cache correctly?

Comment (3nids) [reply .github/workflows/minw64.yml:37]: This looks good!
How long does it take to build the base image? Should we also cache it?

Comment (elpaso) [reply .github/workflows/minw64.yml:37]: > This looks good!
> How long does it take to build the base image? Should we also cache it?

About 5 minutes. We could cache it as well.

Comment (elpaso) [reply .github/workflows/minw64.yml:37]: @3nids we can add the docker image cache later. For now I'm stuck with the last check in this PR, it seems to be blocked:
```
build Expected — Waiting for status to be reported 
```
do you have any clue?

Comment (elpaso) [reply .github/workflows/minw64.yml:37]: I'm not sure the cache is actually working: I've made a commit that just changed the .yml file and the build is taking the usual looong time.

**Comment by manisandro** [comment]
Cool thanks @elpaso!

**Comment by elpaso** [comment]
> Cool thanks @elpaso!

@manisandro Thank **you** for making this possible!

--

@roya0045 please test the artifacts when you have time, there should be Python now!

**Comment by roya0045** [comment]
@elpaso Great news! I've been fiddling with msbuild and I've been getting some weird errors, it's good to know that this exists as a worst-case scenario now to validate processing changes!

**Comment by 3nids** [comment]
It would have been nice to sort the caching before merging this one.
I have created this PR for a rough start, but someone needs to jump in: https://github.com/qgis/QGIS/pull/41734

**Comment by elpaso** [comment]
@3nids happy to jump in, I can take it from #41734

**Comment by agiudiceandrea** [comment]
@elpaso, it seems the "mingw-w64-x86_64-python3-dateutil" package is needed by the "MetaSearch" core plugin:

`WARNING    Python error : Couldn't load plugin 'MetaSearch' due to an error when calling its classFactory() method See message log (Python Error) for more details.`


```
CRITICAL    Traceback (most recent call last):
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 335, in _startPlugin
              plugins[packageName] = package.classFactory(iface)
              File "F:/QGIS/QGIS-Portable/share/qgis/python/plugins\MetaSearch\__init__.py", line 29, in classFactory
              from MetaSearch.plugin import MetaSearchPlugin
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
              File "F:/QGIS/QGIS-Portable/share/qgis/python/plugins\MetaSearch\plugin.py", line 34, in 
              from MetaSearch.dialogs.maindialog import MetaSearchDialog
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
              File "F:/QGIS/QGIS-Portable/share/qgis/python/plugins\MetaSearch\dialogs\maindialog.py", line 50, in 
              from owslib.csw import CatalogueServiceWeb # spellok
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\owslib\csw.py", line 19, in 
              from owslib import fes
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\owslib\fes.py", line 19, in 
              from owslib import util
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\owslib\util.py", line 13, in 
              from dateutil import parser
              File "F:\QGIS\QGIS-Portable\lib\python3.9\site-packages\qgis\utils.py", line 798, in _import
              mod = _builtin_import(name, globals, locals, fromlist, level)
             ModuleNotFoundError: No module named 'dateutil'
```

**Comment by elpaso** [comment]
@agiudiceandrea you can try to add it to the docker file [1] or open a ticket.

[1] https://github.com/qgis/QGIS/blob/master/ms-windows/mingw/qgis3-build-deps-mingw.dockerfile#L30

**Comment by manisandro** [comment]
@agiudiceandrea @elpaso  https://github.com/qgis/QGIS/pull/41761

**Comment by agiudiceandrea** [comment]
See also https://github.com/qgis/QGIS/issues/41855

**Comment by roya0045** [comment]
@elpaso I just got around to test an artifact with all my PR and one on master with a bugfix, and one issue that I encounter is that all geopackage data from my existing project are transferred to tables. I am not sure if this is an issue with the geopackage driver or something else. Otherwise Python works great!

**Comment by agiudiceandrea** [comment]
@roya0045 I think the issue with GeoPackage in MinGW build (all the internal tables are listed like it was a normal SQLite container and the tables with geometry are not recognised as spatial layer) is due to the fact that the MinGW-GDAL library doesn't properly recognize a gpkg file as a GeoPackage.
See https://github.com/qgis/QGIS/issues/41857#issuecomment-812865772 (GeoPackage) and https://github.com/qgis/QGIS/issues/41857#issuecomment-812907178

**Comment by agiudiceandrea** [comment]
I suppose mingw-gdal was not built with the SQLITE_HAS_COLUMN_METADATA=yes configuration option.

**Comment by manisandro** [comment]
The package comes from here [1], indeed `SQLITE_HAS_COLUMN_METADATA` is not set. I'll look into it.

[1] https://src.fedoraproject.org/rpms/mingw-gdal/tree/rawhide

**Comment by roya0045** [comment]
@manisandro Thank you Sandro!

**Comment by roya0045** [comment]
@manisandro libsqlite3 ( as stated here https://gdal.org/drivers/vector/index.html#vector-drivers ) seem to be missing from https://src.fedoraproject.org/rpms/mingw-gdal/blob/rawhide/f/mingw-gdal.spec, I only see libspatialite, which explains why spatialite are fine. Hopefully this helps.

**Comment by agiudiceandrea** [comment]
There is mingw64-sqlite that ships libsqlite3-0.dll

**Comment by manisandro** [comment]
Should be fixed with mingw-gdal-3.2.2-2.fc35

**Comment by roya0045** [comment]
@agiudiceandrea @manisandro Thank you very much for taking the time to look into this! Will any upcoming build be patched or is it safer to force this version in the dependencies?

**Comment by manisandro** [comment]
Give it a day or two for the update to propagate to the rawhide repos, after that any freshly built docker image will contain that version.

**Comment by roya0045** [comment]
@manisandro Perfect, thank you!

**Comment by agiudiceandrea** [comment]
thanks @manisandro I've just tested using libgdal-28.dll from mingw-gdal-3.2.2-2.fc35 and it works! allowing the QGIS MinGW build to correctly menage (import/export) the GPKG format.

**Pull request merged**